lib/events.js

```diff
@@ -1013,125 +1017,172 @@ function eventTargetAgnosticAddListener(emitter, name, listener, flags) {
+        // we stop listening
+        if (error) {
+          const p = PromiseReject(error);
+          // Only the first element errors
+          error = null;
+          return p;
+        }
 
-      // Then we error, if an error happened
-      // This happens one time if at all, because after 'error'
-      // we stop listening
-      if (error) {
-        const p = PromiseReject(error);
-        // Only the first element errors
-        error = null;
-        return p;
-      }
+        // If the emitter has finished, we close the handler
+        if (finished) return closeHandler();
 
-      // If the iterator is finished, resolve to done
-      if (finished) {
-        return PromiseResolve(createIterResult(undefined, true));
+        // Wait until an event happens
+        return new Promise(function(resolve, reject) {
+          promiseReferences.push({ resolve, reject });
+        });
       }
-
-      // Wait until an event happens
-      return new Promise(function(resolve, reject) {
-        unconsumedPromises.push({ resolve, reject });
-      });
+      const value = unconsumedValues.shift();
+      size--;
+      if (paused && size < lowWatermark) {
+        emitter.resume();
+        paused = false;
+      }
+      return value;
     },
 
     return() {
-      eventTargetAgnosticRemoveListener(emitter, event, eventHandler);
-      eventTargetAgnosticRemoveListener(emitter, 'error', errorHandler);
-
-      if (signal) {
-        eventTargetAgnosticRemoveListener(
-          signal,
-          'abort',
-          abortListener,
-          { once: true });
-      }
-
-      finished = true;
-
-      for (const promise of unconsumedPromises) {
-        promise.resolve(createIterResult(undefined, true));
-      }
-
-      return PromiseResolve(createIterResult(undefined, true));
+      return closeHandler();
     },
 
     throw(err) {
       if (!err || !(err instanceof Error)) {
         throw new ERR_INVALID_ARG_TYPE('EventEmitter.AsyncIterator',
                                        'Error', err);
       }
-      error = err;
-      eventTargetAgnosticRemoveListener(emitter, event, eventHandler);
-      eventTargetAgnosticRemoveListener(emitter, 'error', errorHandler);
+      errorHandler(err);
+    },
+    [kWatermarkData]: {
+      /**
+       * The current queue size
+       */
+       get size() {
+        return size;
+      },
+      /**
+       * The low watermark. The emitter is resumed every time size is lower than it
+       */
+      get low() {
+        return lowWatermark;
+      },
+      /**
+       * The high watermark. The emitter is paused every time size is higher than it
+       */
+      get high() {
+        return highWatermark;
+      },
+      /**
+       * It checks wether the emitter is paused by the watermark controller or not
+       */
+      get isPaused() {
+        return paused;
+      }
     },
-
-    [SymbolAsyncIterator]() {
-      return this;
-    }
```

Comment:
This part wasn't needed because this method is inherited from the parent prototype